### PR TITLE
#289 - calculate content assist context type for $ref values based on JSON schema, not regex

### PR DESCRIPTION
**Pull request by tfesenko**
This PR introduces a new `ComponentContextType` which uses JSON schema to retrieve context type for $ref values. 
* It eliminates our public static final regexes
* It also fixes code assist context for schema, as the regex-based version was too relaxed and was using schema contexts for schema examples

### CODE REVIEW

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/JsonReferenceProposalProvider.java [hunk 1]
```diff
@@ -69,10 +71,14 @@ public boolean canProvideProposal(JsonPointer pointer) {
...
      */
     public Collection<Proposal> getProposals(JsonPointer pointer, JsonDocument document, Scope scope) {
-        final ContextType type = contextTypes.get(pointer);
+        final ContextType type = contextTypes.get(document.getModel(), pointer);
         final IFile currentFile = getActiveFile();
         final IPath basePath = currentFile.getParent().getFullPath();
         final List<Proposal> proposals = Lists.newArrayList();
+        
+        TypeDefinition schemaType = document.getModel().find(pointer).getType();
+        System.out.println("SCHEMA TYPE: " + schemaType.getPointer());
```
Comment (ghillairet) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/JsonReferenceProposalProvider.java]: Maybe this log should be removed.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public boolean canProvideProposal(Model model, JsonPointer pointer) {
+        if (model == null) {
+            // model can be null when initTextMessages called in new JsonContentAssistProcessor()
+            return false;
+        }
+        return isReference(model, pointer) && isReferenceToComponent(model, pointer);
+    }
+
+    protected boolean isReference(Model model, JsonPointer pointer) {
+        TypeDefinition type = model.find(pointer).getType();
```
Comment (ghillairet) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java]: Model.find can return null, are you sure it won't in this case?

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        }
+        JsonPointer pointerToType = type.getPointer();
+        if (pointerToType == null) {
+            return false;
+        }
+        return getReferencePointerString().equals(pointerToType.toString());
+    }
+
+    protected boolean isReferenceToComponent(Model model, JsonPointer pointer) {
+        TypeDefinition parentType = model.find(pointer.head()).getType();
```
Comment (ghillairet) [inline com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java]: Same here.

**Review by ghillairet — APPROVED** [review]
Looks good, only some comments about possible null pointer exceptions.

File: com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java (hunk 3 above)
Comment (tfesenko) [reply com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java]: Thanks, @ghillairet , I will address your comments

Comment (tfesenko) [reply com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/assist/contexts/ComponentContextType.java]: @ghillairet , I addressed your comments in ca3f755
I will merge this PR now to get these changes to the integration build. Feel free to re-review this PR

**Pull request merged**